### ISSUE TITLE
Providing OS thread ids when tracing

### ISSUE THREAD

**Issue by atpoverload** [tensorflow/tensorflow#52992]
**System information**

 - Debian 9
 - python 3.5
 - TensorFlow 2.0.0

**Are you willing to contribute it (Yes/No):**

Yes

**Describe the feature and the current behavior/state.**

Currently, tracing only reports internal ids for threads:

```json
{
    "ph": "t",
    "cat": "DataFlow",
    "name": "bert/encoder/layer_0/output/dense/bias",
    "pid": 1,
    "tid": 5,
    "ts": 1635482148893406,
    "id": 1
}
```

It would helpful to be able to also report the OS thread id. Unfortunately, `threading.enumerate` does not appear to report these threads. Examining the runtime from the OS also does not have enough information to identify which is the executing thread.

**Will this change the current api? How?**

I am not quite sure of all the implications but I am imagining a boolean argument to enable collection of the ids:

```python
# hook for estimator
hook = tf.estimator.ProfilerHook(save_secs=1, output_dir=output, os_ids=True)

# tensorboard callback
callback = tf.keras.callbacks.TensorBoard(log_dir = logs,
                                                 histogram_freq = 1,
                                                 profile_batch = '500,520',
                                                 os_ids=True)
```

In addition, downstream concurrent code such as threads, pools, etc would need to changed to grab the id, probably with `getppid`. Finally, the tracing data structures need to be updated to include this field. The new trace output could look something like:

```json
{
    "ph": "t",
    "cat": "DataFlow",
    "name": "bert/encoder/layer_0/output/dense/bias",
    "pid": 1,
    "tid": 5,
    "os_pid": 120311,
    "ts": 1635482148893406,
    "id": 1
}
```

I am not 100% clear on all the consequences but I am very willing to dig into this once I know where to look.

**Who will benefit with this feature?**

This would allow additional context to be attached to the trace. For example, linux reports detailed task information through the `/proc` system, such as the executing cpu. This can then be attached to data that is broken down by cpu.

**Comment by sachinprasadhs** [comment]
Hi, Estimator will be deprecated from Tensorflow 2.12 since it uses Tensorflow V1 behavior.
Could you please migrate from Estimator to Keras API which will enable you to perform the aforementioned tasks, such as [model building](https://www.tensorflow.org/guide/keras/custom_layers_and_models), gradient application, [training](https://www.tensorflow.org/guide/keras/customizing_what_happens_in_fit), evaluation, and prediction.
Refer Migration guide here https://www.tensorflow.org/guide/migrate/migrating_estimator for more details. Thanks!

**Comment by atpoverload** [comment]
Hi, the example I gave above also applied to the Keras API through the tensorboard callback which also does not provide the OS thread ids.

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
